pandas/core/reshape/merge.py

```diff
@@ -1441,9 +1441,13 @@ def _factorize_keys(lk, rk, sort=True):
         rk = rk.values
 
     # if we exactly match in categories, allow us to use codes
+    if isinstance(lk, CategoricalIndex):
+        ldata = lk._data
+    else:
+        ldata = lk
```

Comment:
@TomAugspurger side note, when your Category Dtype enhancement is added, we can remove all of this ``Categorical.is_dtype_equal`` bizness... as ``pandas.core.dtypes.common.is_dtype_equal`` will just work then (as you will have ``__eq__`` defined for the dtypes)